pyfr/solvers/base/elements.py

```diff
@@ -152,6 +152,13 @@ def set_backend(self, backend, nscal_upts):
         self._scal_upts_temp = backend.matrix(inb.ioshape, aliases=aliases,
                                               tags=inb.tags)
 
+        # Allocate required scratch space for artificial viscosity
+        if self.cfg.getfloat('solver-avis', 'amu0', 0.0):
+            self._avis_upts = avis = backend.matrix((nupts, 1, neles), tags={'align'})
+            self._avis_fpts = alloc('avis_fpts', (nfpts, 1, neles))
+            self._avis_upts_temp = backend.matrix(avis.ioshape, aliases=self._avis_fpts,
```

Comment:
Possibly unsafe, nupts is not always less than nfpts.  You'll need an if statement to decide which one to allocate and which one to alias.  Also, this probably does not belong in `set_backend` for `BaseElements`, instead attach it to `BaseAdvectionDiffusionElements`, or `NavierStokesElements`.